### PR TITLE
Add to the JDBC test framework the ability to set a performance SLA per file

### PR DESCRIPTION
**Pull request by basasairohan**
### Description

- Currently the tests developed on Babelfish do not have performance SLAs. This change adds SLAs to the JDBC test framework. 
- Added a new parameter in `config.txt` to specify the default SLA. This will be considered as default timeout unless it is explicitly mentioned in the test file.
- One can specify the SLA value in milli seconds for a specific test file by adding a comment at the starting of the test file mentioning the SLA value. Ex: `-- SLA 10000`

Signed-off-by: Sai Rohan Basa [bsrohan@amazon.com](mailto:bsrohan@amazon.com)

 
### Issues Resolved

[BABEL-3523]

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: test/JDBC/src/main/java/com/sqlsamples/batch_run.java [hunk 1]
```diff
@@ -91,7 +93,7 @@ static void batch_run_sql(Connection con_bbl, BufferedWriter bw, String testFile
                         jdbcPreparedStatement.testPreparedStatementWithFile(result, bw, strLine, logger);
 
                     } else if (!result[1].equals("exec")) {
-
+
```
Comment (thephantomthief) [inline test/JDBC/src/main/java/com/sqlsamples/batch_run.java]: nit: unnecessary spaces

Comment (thephantomthief) [inline test/JDBC/src/main/java/com/sqlsamples/batch_run.java:96]: nit: unnecessary spaces

Comment (thephantomthief) [inline test/JDBC/src/main/java/com/sqlsamples/batch_run.java:96]: nit: unnecessary spaces

File: test/JDBC/src/main/java/com/sqlsamples/batch_run.java [hunk 2]
```diff
@@ -186,7 +188,7 @@ static void batch_run_sql(Connection con_bbl, BufferedWriter bw, String testFile
                     
                 // if line starts with keyword "cursor", it means it is a cursor operation
                 } else if (strLine.startsWith("cursor")) {
-
+
```
Comment (thephantomthief) [inline test/JDBC/src/main/java/com/sqlsamples/batch_run.java]: nit: unnecessary spaces

File: test/JDBC/src/main/java/com/sqlsamples/batch_run.java (hunk 1 above)
Comment (thephantomthief) [inline test/JDBC/src/main/java/com/sqlsamples/batch_run.java]: nit: unnecessary spaces

File: test/JDBC/src/test/java/com/sqlsamples/TestQueryFile.java [hunk 3]
```diff
@@ -396,28 +406,54 @@ public void TestQueryBatch(String inputFileName) throws SQLException, ClassNotFo
...
+        slaMap.add(new AbstractMap.SimpleEntry<>(inputFileName, tempSla)); //add execution time and SLA
+        sla = 0L;
 
         try {
             Assertions.assertTrue(result);
         } catch (AssertionError e) {
-            Throwable throwable = new Throwable(inputFileName + " FAILED! Output diff can be found in '" + diffFile.getAbsolutePath() + "'");
-            throwable.setStackTrace(new StackTraceElement[0]);
-            throw throwable;
+            if(timeout){
```
Comment (thephantomthief) [inline test/JDBC/src/test/java/com/sqlsamples/TestQueryFile.java]: There should be a way to distinguish if a test failed solely due to timeout or if it failed due to both a diff and a timeout. From this logic, it seems that if a test times out, we just say Execution timed out but we do not see if there was potentially a diff as well.

File: test/JDBC/src/test/java/com/sqlsamples/TestQueryFile.java [hunk 4]
```diff
@@ -396,28 +406,54 @@ public void TestQueryBatch(String inputFileName) throws SQLException, ClassNotFo
...
         bw.close();
-        
+        if(sla == 0){
+            sla = defaultSLA*1000000L;
+        }
+        int n = exec_times.size();
+        int index = n - num_lines;
+        long total_time = 0;
+        for(int i = index; i < n; i++){
+            total_time += exec_times.get(i);
```
Comment (thephantomthief) [inline test/JDBC/src/test/java/com/sqlsamples/TestQueryFile.java]: Can't we just sum all the values in exec_times blindly. I didn't quite understand the index logic here

File: test/JDBC/src/main/resources/config.txt [hunk 5]
```diff
@@ -41,3 +41,6 @@ testFileRoot = ./
 
 # WHETHER TEST RUN MODE IS UPGRADE TEST RUN
 isUpgradeTestMode = false
+
+# Default SLA
```
Comment (thephantomthief) [inline test/JDBC/src/main/resources/config.txt]: We should mention that this value is in milliseconds for clarity

File: test/JDBC/src/main/java/com/sqlsamples/batch_run.java [hunk 6]
```diff
@@ -49,6 +52,7 @@ static void batch_run_sql(Connection con_bbl, BufferedWriter bw, String testFile
             FileInputStream fstream;    // stream of input file with the SQL batch queries to be tested
             DataInputStream in;
             BufferedReader br;
+            int file_length = 0;
```
Comment (thephantomthief) [inline test/JDBC/src/main/java/com/sqlsamples/batch_run.java]: Please remove this variable now.

**Review by thephantomthief — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]

**Pull request merged**